userspace/libsinsp/runc.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (C) 2013-2019 Draios Inc dba Sysdig.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include "runc.h"
+
+#include <cstring>
+
+#include "sinsp.h"
+#include "sinsp_int.h"
+
+namespace {
+
+const size_t CONTAINER_ID_LENGTH = 64;
+const size_t REPORTED_CONTAINER_ID_LENGTH = 12;
+const char* CONTAINER_ID_VALID_CHARACTERS = "0123456789abcdefABCDEF";
+
+static_assert(REPORTED_CONTAINER_ID_LENGTH <= CONTAINER_ID_LENGTH, "Reported container ID length cannot be longer than actual length");
+
+bool detect_one_container_id(const std::string& cgroup, const std::string& prefix, const std::string& suffix, std::string& container_id)
+{
+	size_t pos = cgroup.rfind(prefix);
+	if (pos == std::string::npos)
+	{
+		return false;
+	}
+	pos += prefix.size();
+
+	size_t pos2 = cgroup.rfind(suffix);
+	if (pos2 == std::string::npos)
+	{
+		return false;
+	}
+
+	if (pos2 - pos != CONTAINER_ID_LENGTH)
+	{
+		return false;
+	}
+
+	static_assert(std::string::npos == (size_t)(-1), "std::string::npos must be largest valid size_t");
+	size_t invalid_ch_pos = cgroup.find_first_not_of(CONTAINER_ID_VALID_CHARACTERS, pos);
+	if (invalid_ch_pos < CONTAINER_ID_LENGTH)
+	{
+		return false;
+	}
+
+	container_id = cgroup.substr(pos, REPORTED_CONTAINER_ID_LENGTH);
+	return true;
+}
+
+bool detect_container_id(const std::string& cgroup, const libsinsp::runc::cgroup_layout *layout, std::string &container_id)
+{
+	for(size_t i = 0; layout[i].prefix && layout[i].suffix; ++i)
```

Comment:
this way of doing bounds checking is somewhat risky, as it depends on someone putting null pointers at the end of arrays put in.

We should be able to statically determine the size of those arrays with sizeof(X)/sizeof(X[0]) and then that value can be passed in, or even better, we can wrap the array in a struct which stores the size of the array, calculated at compile time

Reply:
`sizeof` won't work across function calls. Given that we have only two of these arrays, hardcoded in the source and not modifiable at runtime, I'm not sure it's a big deal (if you insist, I'd probably go for `std::array`).

Reply:
i don't think the number of instance is the issue as much as the fact that depending on the last entry of an array to be null in order to prevent a segfault is not really idiomatic, and there's no compelling reason to do it that way when safer options exist.

I didn't mean computing sizeof in the function itself, but along with the array
static const int foo = {1,5,7,2,7};
static const foo_length = sizeof(foo) / sizeof(foo[0]);

Then you use foo_length as your loop bounds, as it's clear and safe.

std::array is fine or even vector would work fine here as well